feat: escrow account sweeping

keruch opened the pull request:
https://github.com/dymensionxyz/hyperlane-monorepo/issues/214

keruch committed:
feat: escrow account sweeping

keruch committed:
feat: escrow account sweeping

keruch commented on dymension/libs/kaspa/lib/validator/src/withdraw.rs:
```diff
@@ -190,27 +230,29 @@ async fn validate_messages(
 
 pub fn validate_pskts(
     bundle: &Bundle,
-    messages: &[Vec<HyperlaneMessage>],
+    expected_messages: &[Vec<HyperlaneMessage>],
     hub_anchor: TransactionOutpoint,
-    must_match: MustMatch,
+    escrow_public: EscrowPublic,
+    address_prefix: KaspaAddrPrefix,
 ) -> Result<(), ValidationError> {
-    if bundle.0.len() != messages.len() {
+    if bundle.0.len() != expected_messages.len() {
         return Err(ValidationError::MessageCacheLengthMismatch {
             expected: bundle.0.len(),
-            actual: messages.len(),
+            actual: expected_messages.len(),
         });
     }
 
     // PSKTs must be linked by anchor, starting with the current hub anchor
     let mut anchor_to_spend = hub_anchor;
     for (idx, pskt) in bundle.iter().enumerate() {
-        let messages = messages.get(idx).unwrap();
+        let messages = expected_messages.get(idx).unwrap();
```
there was a bug when `messages` was overwriting itself

keruch commented on dymension/libs/kaspa/lib/validator/src/withdraw.rs:
```diff
@@ -190,27 +230,29 @@ async fn validate_messages(
+    expected_messages: &[Vec<HyperlaneMessage>],
     hub_anchor: TransactionOutpoint,
-    must_match: MustMatch,
+    escrow_public: EscrowPublic,
+    address_prefix: KaspaAddrPrefix,
```
I removed `must_match` and used `escrow_public` + `address_prefix` so I can easily use `validate_pskts` in the demo

keruch commented on dymension/libs/kaspa/lib/validator/src/withdraw.rs:
```diff
@@ -240,22 +308,36 @@ pub fn validate_pskt_impl_details(pskt: &PSKT<Signer>) -> Result<(), ValidationE
+    // Check that the PSKT is a sweeping tx or a withdrawal tx.
+    // If it is a sweeping tx, then the anchor must not be spent.
+    // If it is a withdrawal tx, then the anchor must be spent.
+    match (tx_type, anchor_found) {
+        (TxType::Sweeping, true) => {
+            return Err(ValidationError::AnchorSpent { o: current_anchor });
+        }
+        (TxType::Withdrawal, false) => {
+            return Err(ValidationError::AnchorNotFound { o: current_anchor });
+        }
+        _ => {}
```
Important part!

keruch committed:
feat: escrow account sweeping 1

keruch committed:
cargo fix

keruch committed:
cargo fix

keruch commented on dymension/libs/kaspa/demo/multisig/src/main.rs:
```diff
@@ -126,113 +90,119 @@ async fn load_wallet(args: &Args, url: Option<&str>) -> Result<EasyKaspaWallet>
         .take(e.m())
-        .map(|k| validator_sign_withdrawal_fxg(&safe_b, k, Some(input_selector)))
+        .map(|k| {
+            validate_pskts(
```
I added validation to the demo (without any hub call though)

keruch commented on dymension/libs/kaspa/lib/relayer/src/withdraw/messages.rs:
```diff
@@ -21,13 +28,33 @@ pub async fn on_new_withdrawals(
+    .await
+}
+
+pub async fn build_withdrawal_fxg(
```
I created this function to build WithdrawFXG without any hub calls

keruch committed:
sweeping threshold

keruch committed:
cargo fix

keruch commented on dymension/libs/kaspa/lib/relayer/src/withdraw/hub_to_kaspa.rs:
```diff
@@ -172,10 +156,17 @@ pub fn build_withdrawal_pskt(
+    .map_err(|e| eyre::eyre!("Estimate TX mass: {e}"))?;
+
+    // Apply TX mass multiplier and feerate
+    let tx_fee = (tx_mass as f64 * TX_MASS_MULTIPLIER * feerate).round() as u64;
```
I added feerate, it closes https://github.com/dymensionxyz/hyperlane-monorepo/issues/112

keruch commented on dymension/libs/kaspa/lib/relayer/src/withdraw/hub_to_kaspa.rs:
```diff
@@ -343,45 +342,58 @@ async fn get_utxo_to_spend(
+    Ok((anchor_input, escrow_inputs))
+}
+
+fn estimate_mass(
```
I updated `estimate_fee` (now it is `estimate_mass`) to use a different MassCalculator [which kaspa is using in the tx generator](https://github.com/kaspanet/rusty-kaspa/blob/1adeae8e5e2bdf7b65265420d294a356edc6d9e6/wallet/core/src/tx/generator/test.rs#L202). I found it more accurate and I stopped getting insufficient fee errors.

keruch commented on dymension/libs/kaspa/lib/validator/src/withdraw.rs:
```diff
@@ -221,14 +262,40 @@ pub fn validate_pskt(
-    let ix = validate_pskt_application_semantics(&pskt, must_spend, expected_messages, must_match)?;
-    Ok(TransactionOutpoint::new(pskt.calculate_id(), ix))
+
+    let tx_id = pskt.calculate_id();
+
+    // If there are no messages and payload is empty, then the PSKT
+    // is a sweeping tx which does not spend the anchor
+    let tx_type = if expected_messages.is_empty() && pskt.global.payload.is_none() {
+        info!("PSKT is a sweeping tx: {tx_id}");
+        TxType::Sweeping
+    } else {
+        info!("PSKT is a withdrawal tx: {tx_id}");
+        TxType::Withdrawal
+    };
+
+    let ix = validate_pskt_application_semantics(
+        &pskt,
+        must_spend,
+        tx_type,
+        expected_messages,
+        escrow_public,
+        address_prefix,
+    )?;
+
+    match tx_type {
+        // In case of the sweeping tx, try to spend the anchor in the next PSKT
+        TxType::Sweeping => Ok(must_spend),
+        TxType::Withdrawal => Ok(TransactionOutpoint::new(pskt.calculate_id(), ix)),
+    }
```
important to check!!

keruch commented on dymension/tests/kaspa_hub_test_kas/agent-config.json:
This config worked for me, maybe worth copying from https://github.com/dymensionxyz/hyperlane-deployments/blob/main/playground/guide_kaspa_kas/roles/relayer/agent-config.json

keruch committed:
minors

danwt committed:
fmt

danwt committed:
update claude.md

danwt commented on dymension/libs/kaspa/demo/multisig/src/main.rs:
```diff
@@ -126,113 +90,119 @@ async fn load_wallet(args: &Args, url: Option<&str>) -> Result<EasyKaspaWallet>
-    let args = Args::parse();
+    let mut args = Args::parse();
+
+    args.wallet_secret = Some("123456qwe".to_string());
```
lol

keruch replied:
security is my second name

danwt committed:
fix demo wallet password

danwt commented on dymension/libs/kaspa/lib/core/src/balance.rs:
```diff
@@ -11,12 +11,16 @@ pub async fn check_balance<T: RpcApi + ?Sized>(
-        .get_balance_by_address(addr.clone())
+    let utxos = rpc
+        .get_utxos_by_addresses(vec![addr.clone()])
         .await
-        .map_err(|e| Error::Custom(format!("Getting balance for address: {e}")))?;
+        .map_err(|e| Error::Custom(format!("Getting UTXOs for address: {e}")))?;
+
+    let num = utxos.len();
+    let balance: u64 = utxos.into_iter().map(|u| u.utxo_entry.amount).sum();
+
+    info!("{} has {} UTXOs and {} balance", source, num, balance);
 
-    info!("{} balance: {}", source, balance);
     Ok(balance)
```
@keruch what was wrong with the old way?

keruch replied:
nothing, I just wanted to know the number of UTXOs

danwt committed:
progress checkpoint: code might be broken

danwt committed:
claude

danwt commented on dymension/libs/kaspa/lib/hardcode/src/tx.rs:
```diff
@@ -1,8 +1,23 @@
+// Add this small priproty fee to every sweeping TX to ensure the sufficient fee even if there
+// is one validator. 3_000 is a magic number.
+// TODO: make it configurable?
+pub const RELAYER_SWEEPING_PRIORITY_FEE: u64 = 3_000;
```
is this also going to work on mainnet?

keruch replied:
not sure, that's why I think it's worth making it configurable. 3000 is just what I came up with while testing.

danwt commented on rust/main/hyperlane-base/src/kas_hack/logic_loop.rs:
```diff
@@ -386,9 +388,17 @@ where
 
+        let tx_hash = h512_to_cosmos_hash(outcome.transaction_id).encode_hex_upper::<String>();
+
+        if !outcome.executed {
+            return Err(eyre::eyre!(
+                "Indicate progress failed, TX was not executed on-chain, tx hash: {tx_hash}"
+            ));
+        }
+
         info!(
-            "Dymension, indicated progress on hub: {:?}, outcome: {:?}",
-            fxg.progress_indication, outcome
+            "Dymension, indicated progress on hub: {:?}, outcome: {:?}, tx hash: {:?}",
+            fxg.progress_indication, outcome, tx_hash,
         );
```
🚀

danwt committed:
typo

danwt committed:
comment

danwt committed:
clarification

danwt commented on dymension/libs/kaspa/lib/relayer/src/withdraw/messages.rs:
```diff
@@ -60,15 +150,37 @@ pub async fn on_new_withdrawals(
+    let messages = {
+        let sweep_count = sweeping_bundle.as_ref().map_or(0, |b| b.0.len());
+        let mut messages = Vec::with_capacity(sweep_count + valid_msgs.len());
+        messages.extend(vec![Vec::new(); sweep_count]);
```
what is the `Vec::new()` here for?

keruch replied:
`vec![V; N]` (with semicolon) is a macro to create a vector of `V` repeated `N` times. so with `vec![Vec::new(); sweep_count]` I create a vector of `sweep_count` empty vectors, ie `[[], [], [], ...]`.

danwt replied:
ah i see!

danwt committed:
typos

danwt committed:
fmt

danwt commented on dymension/libs/kaspa/lib/validator/src/withdraw.rs:
```diff
@@ -240,22 +307,36 @@ pub fn validate_pskt_impl_details(pskt: &PSKT<Signer>) -> Result<(), ValidationE
+        _ => {}
     }
 
     let payload_expect = MessageIDs::from(expected_messages).to_bytes();
```
the payload check still works properly? i.e. sweep is checked to be empty?

keruch replied:
yes! it's here 

https://github.com/dymensionxyz/hyperlane-monorepo/blob/78be16551607a6ab32022438b3e6823d3b0cbc07/dymension/libs/kaspa/lib/validator/src/withdraw.rs#L274-L280

danwt commented on dymension/libs/kaspa/lib/validator/src/withdraw.rs:
```diff
@@ -221,14 +262,40 @@ pub fn validate_pskt(
+
+    // If there are no messages and payload is empty, then the PSKT
+    // is a sweeping tx which does not spend the anchor
+    let tx_type = if expected_messages.is_empty() && pskt.global.payload.is_none() {
```
is this clause strictly needed?
> && pskt.global.payload.is_none()
it's confusing me

i think the code will still be correct without this?

keruch replied:
I wanted to distinguish between sweeping and withdrawal TXs in the validation. In case of sweeping, we don't want the anchor to be an input. In case of withdrawal, we do.

https://github.com/dymensionxyz/hyperlane-monorepo/blob/78be16551607a6ab32022438b3e6823d3b0cbc07/dymension/libs/kaspa/lib/validator/src/withdraw.rs#L329-L340

danwt replied:
Right, but I mean, you can check expected messages empty, there is no need to check the payload too in this if statement?
It makes it harder to understand IMO because then there is the payload correctness check in the next function clal

keruch replied:
> the payload correctness check in the next function

now I got it! fixed

danwt reviewed (dismissed):
Great work! Just a few clarifying questions

keruch committed:
threads

danwt committed:
Fix semantic misuse of signature_script field (fixes #227)

- Changed PopulatedInput to include optional redeem_script as third element
- Keep signature_script empty for unsigned transactions (as it should be)
- Store redeem_script separately and only add to PSKT when needed
- Update all usages to handle the new tuple structure

This fixes the semantic confusion where redeem_script was incorrectly
stored in the signature_script field of TransactionInput.

srene commented on dymension/libs/kaspa/lib/validator/src/withdraw.rs:
```diff
@@ -240,22 +307,36 @@ pub fn validate_pskt_impl_details(pskt: &PSKT<Signer>) -> Result<(), ValidationE
     pskt: &PSKT<Signer>,
-    must_spend: TransactionOutpoint,
+    current_anchor: TransactionOutpoint,
+    tx_type: TxType,
```
i think in case of tx_type sweeping it should skip the escrow_inputs_sum != escrow_outputs_sum validation. i tested and i see ValidationError::EscrowAmountMismatch error. i understand that sweeping will never match this comparison because escrow_outputs_sum will be 0 since no hl messages are included.

danwt replied:
it shouldnt happen Sergie because there is the ouput to consolidate
https://github.com/dymensionxyz/hyperlane-monorepo/blob/27ffc02c09fbb133b0da3db77342d9e7d6417bea/dymension/libs/kaspa/lib/validator/src/withdraw.rs#L414
how did you get an error?

srene replied:
my bad issue was actually coming from not being able to validate output pubkey in the loop. probably config issue. solved
